packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+In some environments, the developer can ensure that they remain equivalent as new versions are released.
+For example, consider an application where all the documents are stored on a central server.
+To release a new version, a developer takes the following steps to upgrade from schema A to schema B:
+
+1. At first, the documents are using stored schema A and the application is written against stored schema A.
+2. The developer temporarily takes the service offline, suspending application availability, and then upgrades all documents on the server to use stored schema B before re-enabling the service.
+   Thereafter, any clients attempting to connect to the service will be prompted to download the new version of the application client first, which is written against view schema B.
+3. The migration is complete.
+
+That makes a clean and relatively simple transition.
+The diagram below visualizes the clean break between the old and new schema that happens at step 2:
+
+```mermaid
+gantt
+dateFormat D
+axisFormat Step %e
+tickInterval 1day
+Maintain view schema A in application source code : 1, 1d
+Documents contain stored schema A                 : 1, 1d
+Maintain view schema B in application source code : 2, 1d
+Documents contain stored schema B                 : 2, 1d
+```
+
+Unfortunately, many applications operate under additional constraints.
+For example, what if the developer does not have control over all the documents?
+Perhaps the documents can be stored locally on a user's computer, or can be hosted by another service that is independent of the developer.
+What if the developer does not have control over when a client updates their application?
+Even "live service" applications, like web apps, might allow users to continue using an older/cached version of an application client long after a deployment of a new version.
+In these scenarios, when does the document get upgraded?
+When does the client move to the new version of the application?
+The developer cannot know, and therefore cannot guarantee that the stored schema and view schema of their application will be the same for all clients and documents.
+
+This is problematic.
+If view schema B is deployed to the application, a client might encounter a document that is still using stored schema A (i.e. the view schema is ahead of the stored schema).
+Or, a client might avoid updating even after its document has upgraded to the stored schema B (i.e. the view schema is behind the stored schema).
+The application cannot function unless the view schema and the stored schema agree.
+
+How then does a developer in this situation go about upgrading their application schema?
+The developer must do a **staged rollout** - an incremental set of application deployments that release client code only when it is safe to do so.
+
+1. Initially, the documents are using stored schema A and the application is written against view schema A.
+2. View schema B is written into the application client code.
```

Comment:
I'd summarize this paragraph by saying that r/w support of schema B is shipped darked, but engaged if document is opened in schema B. Then elaborate what that means. I'd not mention views or UI - in many cases changes are abstracted from presentations layers.